### ISSUE TITLE
Set timeout of a socket?

### ISSUE THREAD

**Issue by malmaud** [JuliaLang/ZMQ.jl#87]
Normally I'd use zmq_setsockopt with rcvtimeo if I only want to wait a fixed amount of time for a socket to receive. The implementation of `recv`in this wrapper though is blocking when EAGAIN is returned, which seems to defeat the purpose of setting rcvtimo.

**Comment by stevengj** [comment]
Wouldn't it be more Julian to use a timer task?

**Comment by malmaud** [comment]
Would you mind elaborating? How would I create a task that is willing to wait up to say 1sec to recv from a socket `s`?

**Comment by stevengj** [comment]
Hmm, I guess it is pretty awkward; all the solutions I can come up with involve one task killing another task after a timeout.

**Comment by malmaud** [comment]
Ya, that was my experience as well.

On Mon, Aug 10, 2015 at 7:46 PM, Steven G. Johnson <notifications@github.com

**Comment by malmaud** [comment]
Well i'm not sure this is the most elegant, but something like this would work if you don't mind a blocked task sticking around indefinitely. 

```julia
s = Socket()
c= Channel()
@async put!((c, :received), recv(s))
@async sleep(1); put!(c, (nothing, :timedout))
data, status = take!(c)
```

**Comment by vtjnash** [comment]
the blocked task will finish when the Channel is closed, so it shouldn't really be necessary to use the timeout feature in this way. I'm assuming you want this for some form of heartbeat / status indication? For that, I think you can instead create a worker loop that does all of the handling for the channel and have it tickle a local watchdog counter variable whenever it wakes up.

**Comment by malmaud** [comment]
I did end up using a solution along those lines in Requests.jl. I still think it's kinda confusing for a ZMQ wrapper to allow setting `rvctimeo` but then to not actually timeout, but given that Julia has ways around this it's not a pressing issue for me anymore.

**Comment by stevengj** [comment]
I agree that it would be nice to timeout if you explicitly requested it.

**Comment by dmiljkovic** [comment]
Jon, 

Could you post your solution.

**Comment by malmaud** [comment]
just what I [posted earlier](https://github.com/JuliaLang/ZMQ.jl/issues/87#issuecomment-131153884)

**Comment by dmiljkovic** [comment]
thanks

**Comment by malmaud** [comment]
Once some form of  https://github.com/JuliaLang/julia/issues/13763 gets in, it will be an elegant solution to this general kind of situation.

**Comment by ModestMC** [comment]
This still seems like a solid issue to close out, shoutout to @JamesWrigley for submitting an actual PR.
